File: apple/RimeQuad/Sources/RimeQuadKit/SchemaPreflight.swift
//
//  SchemaPreflight.swift — 部署前先問「檔案到底齊不齊」
//
//  ── 為什麼需要它 ────────────────────────────────────────────────────────
//  `rs_last_error()` 在部署失敗時是**空字串**(coordination.md §4):
//  librime 只回一個 bool,原因寫進 glog,永遠到不了使用者面前。
//  所以「少了一本詞庫」與「方案語法錯」在畫面上長得一模一樣,
//  都是「部署失敗」四個字。
//
//  這一層把最常見的那一種失敗**提前**算出來,而且是在動 schema_list **之前** ——
//  所以它不但講得出是哪一本詞庫不見了,而且根本不需要回滾。
//  回滾仍然留著,但只服務「檔案都在、librime 還是建不起來」那一種說不準的情況。
//
//  ⚠ **這不是一個完整的 librime 設定解析器,也不該是。**
//  它只認得幾個高頻的鍵,而且**只有在檔案確實不存在時才出聲**。
//  誤判會擋下完全正常的套件,那比漏判更糟 —— 漏判只是回到現狀。
//

import Foundation

public struct SchemaPreflight {

    public enum Kind: String, Sendable {
        case dictionary, schema, config, grammar
    }

    public struct Missing: Equatable, Sendable {
        public let kind: Kind
        public let fileName: String
        public let referencedBy: String

        public var message: T {
            switch kind {
            case .dictionary:
                return T("方案「\(referencedBy)」需要詞典 \(fileName),但使用者目錄與隨附目錄裡都沒有。請一併安裝提供這本詞典的套件。",
                         "方案「\(referencedBy)」需要词典 \(fileName),但用户目录与随附目录里都没有。请一并安装提供这本词典的套件。",
                         "Schema \(referencedBy) needs the dictionary \(fileName), which is in neither the user nor the shared directory. Install the package that provides it.")
            case .schema:
                return T("找不到方案檔 \(fileName) —— 這個方案並沒有安裝。",
                         "找不到方案文件 \(fileName) —— 这个方案并没有安装。",
                         "Schema file \(fileName) is missing — this schema is not installed.")
            case .config:
                return T("方案「\(referencedBy)」引用了設定檔 \(fileName),但找不到它。",
                         "方案「\(referencedBy)」引用了配置文件 \(fileName),但找不到它。",
                         "Schema \(referencedBy) refers to the config file \(fileName), which is missing.")
            case .grammar:
                return T("方案「\(referencedBy)」需要語言模型 \(fileName),但找不到它。",
                         "方案「\(referencedBy)」需要语言模型 \(fileName),但找不到它。",
                         "Schema \(referencedBy) needs the language model \(fileName), which is missing.")
            }
        }
    }

    public struct Report: Equatable, Sendable {
        public let schemaId: String
        public let missing: [Missing]
        public var ok: Bool { missing.isEmpty }
    }

    /// - Parameter searchDirs: **必須**照 librime 自己的查找順序:
    ///   `[使用者目錄, 隨附目錄]`。反過來的話,使用者覆寫過的檔案會被
    ///   隨附的那一份遮住,而檢查結果與實際載入的東西不一致。
    public static func check(schemaFile: URL, searchDirs: [URL],
                             fm: FileManager = .default) -> Report {
        let base = schemaFile.lastPathComponent
        let fallbackId = base.hasSuffix(SchemaCatalog.suffix)
            ? String(base.dropLast(SchemaCatalog.suffix.count)) : base
        guard let text = try? String(contentsOf: schemaFile, encoding: .utf8) else {
            return Report(schemaId: fallbackId,
                          missing: [Missing(kind: .schema, fileName: base, referencedBy: base)])
        }
        let header = SchemaCatalog.readHeader(text)
        let declaredId = header.schemaId ?? fallbackId

        var refs: [(String, Kind)] = []
        collect(text: text, into: &refs)

        var missing: [Missing] = []
        var seen = Set<String>()
        for (name, kind) in refs {
            guard !seen.contains(name) else { continue }
            seen.insert(name)
            let exists = searchDirs.contains {
                fm.fileExists(atPath: $0.appendingPathComponent(name).path)
            }
            if !exists {
                missing.append(Missing(kind: kind, fileName: name, referencedBy: declaredId))
            }
        }
        return Report(schemaId: declaredId, missing: missing)
    }

    /// 逐行掃描,依**上一個看到的鍵**解讀值。
    ///
    /// 序列會沿用父鍵,這正是 `dependencies:\n  - stroke` 解得出來的原因。
    ///
    /// ⚠ 這裡不用 MiniYaml,理由與 `SchemaCatalog.readHeader` 相同:
    /// 方案檔是完整 YAML,RTS 解析器會在合法的方案上報錯。
    static func collect(text: String, into out: inout [(String, Kind)]) {
        var currentKey = ""
        for raw in text.components(separatedBy: "\n") {
            var line = raw.replacingOccurrences(of: "\r", with: "")
            // 砍掉行尾註解。引號裡的 `#` 很罕見,而誤砍的代價只是少檢查一項。
            if let hash = line.firstIndex(of: "#"),
               line[line.startIndex..<hash].trimmingCharacters(in: .whitespaces).isEmpty
                || !line[line.startIndex..<hash].contains("\"") {
                line = String(line[line.startIndex..<hash])
            }
            let trimmed = line.trimmingCharacters(in: .whitespaces)
            if trimmed.isEmpty { continue }

            if trimmed.hasPrefix("- ") || trimmed == "-" {
                let value = SchemaCatalog.stripQuotes(
                    String(trimmed.dropFirst()).trimmingCharacters(in: .whitespaces))
                add(key: currentKey, value: value, into: &out)
                continue
            }
            guard let colon = trimmed.firstIndex(of: ":") else { continue }
            let key = String(trimmed[trimmed.startIndex..<colon]).trimmingCharacters(in: .whitespaces)
            let value = SchemaCatalog.stripQuotes(
                String(trimmed[trimmed.index(after: colon)...]).trimmingCharacters(in: .whitespaces))
            currentKey = key
            if !value.isEmpty, !value.hasPrefix("[") { add(key: key, value: value, into: &out) }
            // 行內序列 `dependencies: [stroke, terra_pinyin]`
            if value.hasPrefix("["), value.hasSuffix("]") {
                for item in value.dropFirst().dropLast().components(separatedBy: ",") {
                    add(key: key, value: SchemaCatalog.stripQuotes(
                        item.trimmingCharacters(in: .whitespaces)), into: &out)
                }
            }
        }
    }

    static func add(key: String, value: String, into out: inout [(String, Kind)]) {
        guard !value.isEmpty else { return }
        switch key {
        case "dictionary":
            out.append(("\(value).dict.yaml", .dictionary))
        case "dependencies":
            out.append(("\(value)\(SchemaCatalog.suffix)", .schema))
        case "import_preset":
            out.append(("\(value).yaml", .config))
        case "__include", "__patch":
            if let target = includeTarget(value) { out.append((target, .config)) }
        case "language":
            // librime 的語言模型檔名慣例是 `zh-hans-t-essay-bgw.gram` 這種帶連字號的。
            // 不含連字號的 `language:` 多半是別的意思,不要猜。
            if value.contains("-") { out.append(("\(value).gram", .grammar)) }
        default:
            break
        }
    }

    /// `pinyin:/xxx` → `pinyin.yaml`;`symbols.yaml:/xxx` → `symbols.yaml`;
    /// `/local/node` → nil(同文件內的節點參照,不需要檔案)。
    static func includeTarget(_ raw: String) -> String? {
        let v = raw.trimmingCharacters(in: .whitespaces)
        guard !v.hasPrefix("/") else { return nil }
        var head = v.components(separatedBy: ":/")[0]
        head = head.components(separatedBy: ":")[0].trimmingCharacters(in: .whitespaces)
        guard !head.isEmpty else { return nil }
        return head.hasSuffix(".yaml") ? head : head + ".yaml"
    }

    /// 一組方案的預檢。任何一個有問題就整組不通過 ——
    /// 部署是全有全無的,放行一半沒有意義。
    public static func checkAll(schemaIds: [String], searchDirs: [URL],
                                fm: FileManager = .default) -> [Missing] {
        var out: [Missing] = []
        for id in schemaIds {
            let name = id + SchemaCatalog.suffix
            guard let dir = searchDirs.first(where: {
                fm.fileExists(atPath: $0.appendingPathComponent(name).path)
            }) else {
                out.append(Missing(kind: .schema, fileName: name, referencedBy: id))
                continue
            }
            out.append(contentsOf: check(schemaFile: dir.appendingPathComponent(name),
                                         searchDirs: searchDirs, fm: fm).missing)
        }
        return out
    }
}

File: android/app/src/main/java/org/luminakey/ime/store/SchemaPreflight.kt
package org.luminakey.ime.store

import org.luminakey.ime.theme.MiniYaml
import org.luminakey.ime.theme.YamlNode
import java.io.File

/**
 * 部署前的相依檢查。
 *
 * 規範 §4 要求：使用者自帶檔案部署失敗時，「錯誤訊息必須明確告訴使用者
 * **缺少哪一個詞典**，而不是只說部署失敗」。
 *
 * 而 `rs_last_error()` 給不了這個資訊 —— librime 的部署失敗只會回一個
 * 布林，細節散在 glog 的輸出裡，行動端拿不到。所以與其事後解讀，
 * 不如**事前自己算**：把 schema yaml 讀一遍，把它宣告要用的檔案列出來，
 * 逐一確認在 user / shared 目錄裡找得到。
 *
 * 這樣做還有一個附帶好處：檢查在**修改 schema_list 之前**跑，
 * 缺東西時根本不會去動 default.custom.yaml，也就不需要回滾。
 * 回滾仍然必須存在（見 [SchemaStore]），但它只需要處理
 * 「檔案都在、librime 仍然編不起來」這種預測不到的情形。
 *
 * ⚠ 這**不是**一個完整的 librime 配置解析器。它刻意只認幾個高頻欄位，
 * 而且只在「找不到」時說話 —— 誤報會擋住合法套件，比漏報更糟。
 *
 * ## 為什麼要分 [Severity]
 *
 * 第一版把每一個找不到的檔案都當成「不給啟用」，結果**市集 98 個方案裡
 * 有 20 個被自己的預檢擋死**（倉頡系列的 `luna_quanpin`、粵拼與魔然的
 * `cangjie5`、五筆與鄭碼的 `pinyin_simp`……）—— 而 librime 根本不在乎
 * 這些檔案在不在。查 librime 原始碼確認過分界線：
 *
 * | 宣告 | librime 的行為 | 我們 |
 * |---|---|---|
 * | `translator/dictionary` 缺 | `SchemaUpdate::Run` 回 false，**部署真的失敗** | 擋 |
 * | `__include` / `import_preset` 的目標缺 | 配置編譯器 `resource not found` → 回 false | 擋 |
 * | `schema/dependencies` 缺 | `WorkspaceUpdate`：`skipped unsatisfied dependency`，只是 warning | 放行 |
 * | `reverse_lookup/dictionary` 等次要詞典缺 | `SchemaUpdate` 只看 `translator/dictionary`，碰都沒碰 | 放行 |
 * | `grammar/language` 的 `.gram` 缺 | 語言模型建不起來，翻譯器照常運作 | 放行 |
 *
 * 放行的那些仍然照樣回報，只是標成 [Severity.WARNING]：使用者該知道
 * 「這個方案的筆畫反查用不了」，但那不是不讓他打字的理由。
 */
object SchemaPreflight {

    enum class Kind {
        /** translator / reverse_lookup 的 dictionary。這是最常缺的東西。 */
        DICTIONARY,

        /** schema/dependencies 指名的另一個方案。 */
        SCHEMA,

        /** __include / import_preset 指向的配置檔。 */
        CONFIG,

        /** grammar/language 指向的語言模型。 */
        GRAMMAR,
    }

    /** 缺這個檔案會不會讓 librime 的部署整個失敗。 */
    enum class Severity {
        /** 部署一定失敗 —— 不要動 schema_list。 */
        BLOCKING,

        /** 部署會成功，但某個功能不會有作用。告知，不阻擋。 */
        WARNING,
    }

    data class Missing(
        val kind: Kind,
        val fileName: String,
        val referencedBy: String,
        val severity: Severity = Severity.BLOCKING,
    ) {
        fun humanMessage(): String = when (severity) {
            Severity.BLOCKING -> when (kind) {
                Kind.DICTIONARY ->
                    "方案「$referencedBy」需要詞典 $fileName，但 user 與 shared 目錄裡都沒有。" +
                        "請一併導入提供這本詞典的套件。"
                Kind.SCHEMA ->
                    "方案「$referencedBy」宣告相依於方案檔 $fileName，目前找不到。"
                Kind.CONFIG ->
                    "方案「$referencedBy」引用了配置檔 $fileName，目前找不到。"
                Kind.GRAMMAR ->
                    "方案「$referencedBy」需要語言模型 $fileName，目前找不到。"
            }
            // 措辭刻意跟上面不同：使用者要能一眼看出「這條不擋我」。
            Severity.WARNING -> when (kind) {
                Kind.DICTIONARY ->
                    "方案「$referencedBy」的次要詞典 $fileName 不在，可以照常打字，" +
                        "但用到這本詞典的功能（多半是反查、輔助碼）不會有作用。"
                Kind.SCHEMA ->
                    "方案「$referencedBy」宣告相依於方案 $fileName，目前沒安裝；" +
                        "librime 會略過它，主方案照常可用。"
                Kind.CONFIG ->
                    "方案「$referencedBy」引用的配置檔 $fileName 不在。"
                Kind.GRAMMAR ->
                    "方案「$referencedBy」的語言模型 $fileName 不在，整句輸入的排序會差一些，" +
                        "其餘照常。"
            }
        }
    }

    data class Report(val schemaId: String, val missing: List<Missing>) {
        val blocking: List<Missing> get() = missing.filter { it.severity == Severity.BLOCKING }
        val warnings: List<Missing> get() = missing.filter { it.severity == Severity.WARNING }

        /** 「可以啟用嗎」。注意這**不是**「沒有任何東西缺」。 */
        val ok: Boolean get() = blocking.isEmpty()
    }

    /**
     * 檢查 [schemaFile]（一份 `*.schema.yaml`）。
     * [searchDirs] 依 librime 的搜尋順序給：user 在前、shared 在後。
     */
    fun check(schemaFile: File, searchDirs: List<File>): Report {
        val text = runCatching { schemaFile.readText(Charsets.UTF_8) }.getOrNull()
            ?: return Report(
                schemaFile.name,
                listOf(Missing(Kind.SCHEMA, schemaFile.name, schemaFile.name)),
            )
        val id = schemaFile.name.removeSuffix(".schema.yaml")
        return Report(id, checkText(id, text, searchDirs))
    }

    internal fun checkText(schemaId: String, text: String, searchDirs: List<File>): List<Missing> {
        val root = runCatching { MiniYaml.parse(schemaId, text).root }.getOrNull()
            ?: return emptyList() // 解析不了就別亂猜；讓 librime 自己去報錯
        val wanted = LinkedHashMap<String, Pair<Kind, Severity>>()
        collect(root, key = null, topKey = null, out = wanted)

        val declaredId = ((root as? YamlNode.Mapping)
            ?.entries?.get("schema") as? YamlNode.Mapping)
            ?.entries?.get("schema_id")
            ?.let { (it as? YamlNode.Scalar)?.value } ?: schemaId

        return wanted.entries
            .filterNot { (name, _) -> exists(name, searchDirs) }
            .map { (name, kv) -> Missing(kv.first, name, declaredId, kv.second) }
    }

    private fun exists(name: String, dirs: List<File>): Boolean =
        dirs.any { File(it, name).isFile }

    /**
     * 遞迴走整棵樹，收集「這份 yaml 需要哪些檔案」。
     *
     * [topKey] 是「從根往下走時經過的第一層鍵」。分界線就靠它：librime 的
     * `SchemaUpdate::Run` 只編 `translator/dictionary` 這一本，別處的
     * `dictionary:`（`reverse_lookup`、`custom_phrase`、`translator@xxx`……）
     * 它連讀都沒讀。
     */
    private fun collect(
        node: YamlNode,
        key: String?,
        topKey: String?,
        out: MutableMap<String, Pair<Kind, Severity>>,
    ) {
        fun put(name: String, kind: Kind, severity: Severity) {
            val old = out[name]
            // 同一個檔案被兩處引用時，取比較嚴的那一個。
            if (old == null || (old.second == Severity.WARNING && severity == Severity.BLOCKING)) {
                out[name] = kind to severity
            }
        }

        when (node) {
            is YamlNode.Scalar -> {
                val v = node.value ?: return
                if (v.isEmpty()) return
                when (key) {
                    // translator/dictionary 是主詞典，缺了部署一定失敗。
                    // 其餘位置的 dictionary 是次要詞典，librime 不在部署期碰它。
                    "dictionary" -> put(
                        "$v.dict.yaml",
                        Kind.DICTIONARY,
                        if (topKey == "translator") Severity.BLOCKING else Severity.WARNING,
                    )
                    // punctuator/import_preset: default → default.yaml
                    "import_preset" -> put("$v.yaml", Kind.CONFIG, Severity.BLOCKING)
                    // __include: pinyin:/xxx 或 __include: symbols.yaml:/xxx
                    "__include", "__patch" -> includeTarget(v)?.let {
                        put(
                            it.fileName,
                            Kind.CONFIG,
                            // 結尾的 `?` 是 librime 的「可有可無」記號：
                            // ResolveReference 找不到時只印 INFO，Resolve 回 true。
                            if (it.optional) Severity.WARNING else Severity.BLOCKING,
                        )
                    }
                    // grammar/language: zh-hans-t-essay-bgw → .gram
                    "language" -> if (v.contains('-')) {
                        put("$v.gram", Kind.GRAMMAR, Severity.WARNING)
                    }
                    // schema/dependencies 的每一項。librime 找不到時只印 warning
                    // 「skipped unsatisfied dependency」，主方案照樣部署成功。
                    "dependencies" -> put("$v.schema.yaml", Kind.SCHEMA, Severity.WARNING)
                }
            }

            is YamlNode.Sequence ->
                // 序列項目沿用父層的鍵，這樣 dependencies: [- stroke] 才認得出來
                node.items.forEach { collect(it, key, topKey, out) }

            is YamlNode.Mapping ->
                node.entries.forEach { (k, v) -> collect(v, k, topKey ?: k, out) }
        }
    }

    /** [includeTarget] 的結果。 */
    internal data class IncludeRef(val fileName: String, val optional: Boolean)

    /**
     * `__include` 指到哪個檔案，或者根本不指到檔案。
     *
     * ⚠ 這裡曾經整段寫錯，代價是市集裡十個方案按下「啟用」被擋死。
     * librime 自己的註解（`config_compiler.cc`，`ParseInclude` 上方）就是答案：
     *
     * ```
     * // __include: path/to/local/node
     * // __include: filename[.yaml]:/path/to/external/node
     * ```
     *
     * **有沒有冒號才是分界線，不是開頭的斜線。** `CreateReference()` 找不到 `:`
     * 時，resource_id 直接取 `current_resource_id()` —— 也就是同一個檔案。所以
     * 倉頡的 `__include: array30_format`、魔然的 `__include: reverse_format`
     * 都是同檔案裡的節點（那兩份 yaml 各自都有這個頂層鍵），一個檔案都不缺。
     *
     * 結尾的 `?` 代表可有可無（`Reference::optional`），缺了不算失敗。
     */
    internal fun includeTarget(raw: String): IncludeRef? {
        var v = raw.trim()
        val optional = v.endsWith("?")
        if (optional) v = v.dropLast(1).trim()
        val sep = v.indexOf(':')
        if (sep <= 0) return null           // 沒有冒號 → 同檔案內的節點；冒號在開頭 → 同上
        val head = v.substring(0, sep).trim()
        if (head.isEmpty()) return null
        return IncludeRef(if (head.endsWith(".yaml")) head else "$head.yaml", optional)
    }
}
